// yeepay-realname/lib/yeepay-realname.js
var utf8 = require('utf8');
var request = require('request');
var config = require('./config');
var digest = require('./digest');
var p0_Cmd = "FastRealNameAuth";//接口指令，常量。
var pattern = "1";//认证模式为“1”表示只认证信息真实性,常量。
var inNeedParamList = ['p0_Cmd','customerId','name','idCardNumber','bankCode','bankCardNumber','pattern'];
function extend(target,source){
	target = target || {};
	source = source || {};
	for(var key in source){
		if(source.hasOwnProperty(key)){
			target[key] = source[key];
		}
	}
	return target;
}
function toString(text){
	return text.toString();
}
function noop(){};
/**
*检查易宝要求的参数是否齐全，只有齐全了才会向易宝发起实名认证，节约成本
*@param object json
*return boolean
*/
function checkParam(json){
	json = json || {};
	var flag = true;
	for(var i = 0 ,len = inNeedParamList.length; i < len ; i++){
		if(!json[inNeedParamList[i]]){
			flag = false;
			break;
		}
	}
	return flag;
}
function YeepayRealName(json){
	this.config = extend(config,json);
}
/**
*根据提交的数据，获取MD5签名
*@param object jsonData
*return string 
*/
YeepayRealName.prototype.getReqMd5Hmac = function(jsonData){
	jsonData = extend(jsonData,this.getConstans(jsonData));
	if(!checkParam(jsonData)){
		return "";
	}
	var str = "";
	str += jsonData.p0_Cmd;
	str += jsonData.customerId;
	str += jsonData.name;
	str += jsonData.idCardNumber;
	str += jsonData.bankCode;
	str += jsonData.bankCardNumber;
	jsonData.bankName ? (str += jsonData.bankName) : (str += "");
	jsonData.province ? (str += jsonData.province) : (str += "");
	jsonData.city ? (str += jsonData.city) : (str += "");
	str += jsonData.pattern;
	jsonData.res_desc ? (str += jsonData.res_desc) : (str += "");
	return digest.hmacSign(utf8.encode(str),this.config.keyValue);//中文需要转化为utf8编码
}
/**
*获取配置中的常量
*@param object data
*return object
*/
YeepayRealName.prototype.getConstans = function(data){
	return {
		p0_Cmd:data.p0_Cmd || p0_Cmd,
		pattern:data.pattern || pattern,
		customerId:data.customerId || this.config.customerId
	}
}
/**
*去易宝进行实名认证
*/
YeepayRealName.prototype.authorize = function(data,cbf){
	data = data || {};
	cbf = cbf || noop;
	data = extend(data,this.getConstans(data));
	var hmac = this.getReqMd5Hmac(data);
	var self = this;
	if(!hmac){
		return cbf({code:-1,msg:"参数不正确，请检查！"});
	}
	if(!data.busId){
		return cbf({code:-2,msg:"参数不正确，缺少商户订单号(busId)！"});
	}	
	var url = this.getAuthorizationURL();
	var formData = {
		p0_Cmd:toString(data.p0_Cmd),
		customerId:toString(data.customerId),
		name:toString(data.name),
		idCardNumber:toString(data.idCardNumber),
		bankCode:toString(data.bankCode),
		bankCardNumber:toString(data.bankCardNumber),
		bankName:toString(data.bankName || ""),
		province:toString(data.province || ""),
		city:toString(data.city || ""),
		pattern:toString(data.pattern),
		res_desc:toString(data.res_desc || ""),
		busId:toString(data.busId),
		hmac:hmac
	};
	console.log('url:'+url);
	console.log('form:',formData);
	request.post({url:url, form: formData}, function(err,httpResponse,body){
		if(err){
			cbf({code:-3,msg:'请求易宝实名认证接口失败了。',err:err});
		}else{
			if(httpResponse.statusCode == 200){
				try{
					var body = JSON.parse(body);
					cbf(null,{code:0,msg:'success',data:body});
				}catch(e){
					//todo self.verify(body);
					//body:'r1_Code=-1\nerrorMsg=InvalidParam : hmac=c1f48dfd55c556d760fad8a8e7a691a0'
					cbf({code:-4,msg:"易宝数据返回异常，不是json格式",data:body});
				}			
			}else{
				cbf({code:-5,msg:'请求易宝实名认证接口失败了，statusCode:'+httpResponse.statusCode});
			}	
		}
	});
}
/**
*根据易宝返回的结果进行数据合法性认证，防止被篡改
*@param object jsonData
*return boolean
*/
YeepayRealName.prototype.verify = function(jsonData){
	jsonData = jsonData || {};
	if(!jsonData.hmac){
		return false;
	}
	var hmac = jsonData.hmac;
	delete jsonData.hmac;
	delete jsonData.authMsg;
	var str = "";
	str += jsonData.command;
	str += jsonData.customerId;
	str += jsonData.flowId;
	str += jsonData.responseCode;
	str += jsonData.authStatus;
	str += jsonData.pattern;
	str += jsonData.auth_Amt;
	str += jsonData.res_desc;
	var verifyHmac = digest.hmacSign(utf8.encode(str),this.config.keyValue);
	if(hmac === verifyHmac){
		return true;
	}else{
		return false;
	}
}
/**
*获取获取易宝实名认证的地址
*return string
*/
YeepayRealName.prototype.getAuthorizationURL = function(){
	return this.config.yeepayCommonReqURL;
}
module.exports = YeepayRealName
